bson/primitive/primitive_test.go

```diff
@@ -199,3 +200,334 @@ func TestDateTime(t *testing.T) {
 		})
 	})
 }
+
+func TestMarshalingD(t *testing.T) {
+	t.Parallel()
+
+	testcases := []struct {
+		name     string
+		test     D
+		expected M
+	}{
+		{
+			"nil",
+			nil,
+			nil,
+		},
+		{
+			"empty",
+			D{},
+			M{},
+		},
+		{
+			"non-empty",
+			D{
+				{"a", 42},
+				{"b", true},
+				{"c", "answer"},
+				{"d", nil},
+				{"e", 2.71828},
+				{"f", A{42, true, "answer", nil, 2.71828}},
+				{"g", D{{"foo", "bar"}}},
+			},
+			M{
+				"a": 42,
+				"b": true,
+				"c": "answer",
+				"d": nil,
+				"e": 2.71828,
+				"f": A{42, true, "answer", nil, 2.71828},
+				"g": D{{"foo", "bar"}},
+			},
+		},
+	}
+	for _, tc := range testcases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			got, err := json.Marshal(tc.test)
+			assert.NoError(t, err)
+			want, _ := json.Marshal(tc.expected)
+			assert.Equal(t, want, got)
+		})
+	}
+	for _, tc := range testcases {
+		tc := tc
+		t.Run(tc.name+" Indent", func(t *testing.T) {
+			t.Parallel()
+
+			got, err := json.MarshalIndent(tc.test, "<prefix>", "<indent>")
+			assert.NoError(t, err)
+			want, _ := json.MarshalIndent(tc.expected, "<prefix>", "<indent>")
+			assert.Equal(t, want, got)
+		})
+	}
+}
+
+func TestUnmarshalingA(t *testing.T) {
```

Comment:
Optional: Consider a more descriptive test name (borrowed from testable examples name conventions).

```suggestion
func TestA_UnmarshalJSON(t *testing.T) {
```